OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,220 @@
+            return Collections.emptyList();
+        }
+
+        Timber.d("Web Key Directory import: " + name + " using Proxy: " + proxy.getProxy());
+        List<ImportKeysListEntry> entry = new ArrayList<>(1);
+
+        byte[] data = query(webKeyDirectoryURL, proxy);
+        // if we're here that means key retrieval succeeded,
+        // would have thrown an exception otherwise
+        try {
+            try {
+                entry.add(getEntry(data));
+            } catch (UnsupportedOperationException e) {
+                Timber.e("Parsing retrieved Web Key Directory key failed!");
+            }
+        } catch (PgpGeneralException | IOException e) {
+            Timber.e(e, "Failed parsing key from Web Key Directory during search");
+            throw new QueryFailedException("No valid key found on Web Key Directory");
+        }
+        return entry;
+    }
+
+    @Override
+    public String get(String name, ParcelableProxy proxy) throws QueryFailedException {
+        throw new UnsupportedOperationException("Returning armored key from Web Key Directory not supported");
+    }
+
+    private byte[] query(URL url, ParcelableProxy proxy) throws QueryFailedException {
+        try {
+            Timber.d("fetching from Web Key Directory with: " + url + " proxy: " + proxy.getProxy());
+
+            Request request = new Request.Builder()
+                    .url(url)
+                    .build();
+
+            OkHttpClient client = OkHttpClientFactory.getClientPinnedIfAvailable(url, proxy.getProxy());
+            Response response = client.newCall(request).execute();
+
+            if (response.isSuccessful()) {
+                return response.body().bytes();
+            } else {
+                // probably a 404 indicating that the key does not exist
+                throw new QueryFailedException("key not found in Web Key Directory." +
+                        "response:" + response);
+            }
+
+        } catch (IOException e) {
+            Timber.e(e, "IOException at Web Key Directory key download");
+            throw new QueryFailedException("Cannot connect to Web Key Directory. "
+                    + "Check your Internet connection!"
+                    + (proxy.getProxy() == Proxy.NO_PROXY ? "" : " Using proxy " + proxy.getProxy()));
+        }
+    }
+
+    @Override
+    public void add(String armoredKey, ParcelableProxy proxy) throws AddKeyException {
+        throw new UnsupportedOperationException("Uploading keys to Web Key Directory not supported");
+    }
+
+    /**
+     * Web Key Directory returns the entire key even during our searching phase.
+     *
+     * @throws UnsupportedOperationException if the key could not be parsed
+     */
+    @NonNull
+    private static ImportKeysListEntry getEntry(byte[] data)
+            throws UnsupportedOperationException, IOException, PgpGeneralException {
+        ImportKeysListEntry entry = new ImportKeysListEntry();
```

Comment:
why not construct from the uncached key ring entry? I'm not sure if it does everything you need, but a `ImportKeysListEntry.from(UncachedKeyRing)` method seems more appropriate here